client-metrics/src/main/scala/org/http4s/client/metrics/core/Metrics.scala

```diff
@@ -0,0 +1,62 @@
+package org.http4s.client.metrics.core
+
+import cats.effect.{Clock, Resource, Sync}
+import cats.implicits._
+import java.util.concurrent.TimeUnit
+import org.http4s.{Request, Response}
+import org.http4s.client.Client
+import scala.concurrent.TimeoutException
+
+/**
+  * Client middleware to record metrics for the http4s client.
+  *
+  * This middleware will record:
+  * - Number of active requests
+  * - Time duration to receive the response headers, taking into account the HTTP status code
+  * - Total duration of each request, including consumption of the response body, taking into account the HTTP status code
+  * - Number of timeouts
+  * - Number of other errors
+  *
+  * This middleware can be extended to support any metrics ecosystem by implementing the [[MetricsOps]] type
+  */
+object Metrics {
+
+  /**
+    * Wraps a [[Client]] with a middleware capable of recording metrics
+    *
+    * @param ops a algebra describing the metrics operations
+    * @param classifier a function that allows to add a classifier that can be customized per request
+    * @param client the [[Client]] to gather metrics from
+    * @return the metrics middleware wrapping the [[Client]]
+    */
+  def apply[F[_]](ops: MetricsOps[F], classifier: Request[F] => Option[String] = { _: Request[F] =>
+    None
+  })(client: Client[F])(implicit F: Sync[F], clock: Clock[F]): Client[F] = {
+
+    def withMetrics(req: Request[F]): Resource[F, Response[F]] =
+      (for {
+        start <- Resource.liftF(clock.monotonic(TimeUnit.NANOSECONDS))
```

Comment:
I'm a little worried about overhead of `liftF` and `flatMap` on something that runs on every request.  It's probably insignificant, but we could probably optimize this later by fusing some of the `F`s before lifting into something heavier.  I wouldn't sweat it for this PR, because the API is at least correct.

Reply:
Humm. I didn't about the overhead, sorry for that.
At the beginning I was lifting the Resource into the F context but then I was having problems intercepting the errors inside the Resource. But I can try to give another thought if it's important.

Reply:
fs2-1.0.0 is almost out, and this is a breaking change.  I'd rather get this into our next milestone, and optimize later if we're so inclined.  Let's focus on the other issues.

Reply:
ok, let's do that